lib/util.ml

```diff
@@ -28,32 +29,94 @@ let sign_string_sha256 ~key ~basestring =
+    match get_build_number ~context ~build_url with
+    | Ok n -> n
+    | Error msg -> failwith msg
+
   let is_failed_build (n : Github_t.status_notification) =
     n.state = Failure && Re2.matches buildkite_is_failed_re (Option.default "" n.description)
 
-  let new_failed_steps (n : Github_t.status_notification) (repo_state : State_t.repo_state) pipeline =
-    let to_failed_steps branch step statuses acc =
-      (* check if step of an allowed pipeline *)
-      match step with
-      | step when step = pipeline -> acc
-      | step when not @@ Devkit.Stre.starts_with step pipeline -> acc
-      | _ ->
-      match Common.StringMap.find_opt branch statuses with
-      | Some (s : State_t.build_status) when s.status = Failure ->
-        (match s.current_failed_commit, s.original_failed_commit with
-        | Some _, _ ->
-          (* if we have a value for current_failed_commit, this step was already failed and notified *)
-          acc
-        | None, Some { build_link = Some build_link; sha; _ } when sha = n.commit.sha ->
-          (* we need to check the value of the commit sha to avoid false positives *)
-          (step, build_link) :: acc
-        | _ -> acc)
-      | _ -> acc
-    in
-    match n.state = Failure, n.branches with
-    | false, _ -> []
-    | true, [ branch ] -> Common.StringMap.fold (to_failed_steps branch.name) repo_state.pipeline_statuses []
-    | true, _ -> []
+  let new_failed_steps (n : Github_t.status_notification) (repo_state : State_t.repo_state) =
+    if not (is_failed_build n) then failwith "Error: new_failed_steps fn must be called on a finished build";
+    match n.target_url with
+    | None ->
+      (* if we don't have a target_url value, we don't have a build number and cant't track build state *)
+      []
+    | Some build_url ->
+      let { pipeline_name; _ } = parse_context_exn ~context:n.context ~build_url in
+      (match n.state = Failure, n.branches with
+      | false, _ -> []
+      | true, [ branch ] ->
+        (match StringMap.find_opt pipeline_name repo_state.pipeline_statuses with
+        | Some branches_statuses ->
+          (match StringMap.find_opt branch.name branches_statuses with
+          | Some builds_maps ->
+            let current_build_number = get_build_number_exn ~context:n.context ~build_url in
+            let to_previous_failed_steps n build_number (build_status : State_t.build_status) acc =
+              match int_of_string build_number >= n with
+              | true -> acc
+              | false -> build_status.failed_steps @ acc
+            in
+            let previous_failed_steps =
+              StringMap.fold (to_previous_failed_steps @@ int_of_string current_build_number) builds_maps []
+              |> List.sort_uniq Stdlib.compare
+            in
+            let current_build =
+              (* When this function is called, the current build is finished *)
+              Option.get @@ StringMap.find_opt current_build_number builds_maps
```

Comment:
then the comment should be part of the exception raised when it's None

Reply:
This is already the case above [at the beginning](https://github.com/ahrefs/monorobot/blob/master/lib/util.ml#L87) of the `new_failed_steps` function.

The comment is only here to assure the reader that we are not getting an exception here. i will update the comment to make it clearer

Reply:
then maybe just `StringMap.find` with comment